### PR TITLE
cmake: locate files WRT to project home directory

### PR DESCRIPTION
**Pull request by lorc**
Build system can't find file_to_c.py and certificate files,
because CMakeLists.txt have an assumption about those files
locations. So, when you try to use separate build directory,
relative path in CMakeLists.txt are not working as expected.

It is better to use ${CMAKE_HOME_DIRECTORY} instead of assuming
that mentioned files can be reached from build directory.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by lorc** [comment]
Added tag

**Comment by etienne-lms** [comment]
Hi @lorc, I've tested this change but cmake fails with:
```
make[2]: *** No rule to make target `cert/my.csr', needed by `optee_test/host/xtest/regression_8100_my_csr.h'.  Stop.
make[1]: *** [optee_test/host/xtest/CMakeFiles/xtest.dir/all] Error 2
```

I use cmake version 3.12.2.
Did I miss something?

**Comment by lorc** [comment]
Hi @etienne-lms,

I use cmake 3.13.1, but I don't think that this makes any difference. 

There is how I tried build last time:

```
# mkdir build
# cd build
# cmake -DCMAKE_BUILD_TYPE=Release -DOPTEE_TEST_SDK=/home/lorc/mnt/ssd/tee_test_xen/optee_os/out/arm-plat-rcar/export-ta_arm64/ ..
# make
````

It is working just fine:

```
lorc@EPUAKYIW2556 ~/work/optee_test/build (git)-[remotes/lorc/cmake_fix] % make
[  3%] Generating regression_8100_my_csr.h
[  6%] Generating regression_8100_ca_crt.h
[ 10%] Generating regression_8100_mid_crt.h
[ 13%] Generating regression_8100_my_crt.h
Scanning dependencies of target xtest
[ 16%] Building C object host/xtest/CMakeFiles/xtest.dir/adbg/src/adbg_case.c.o
[ 20%] Building C object host/xtest/CMakeFiles/xtest.dir/adbg/src/adbg_enum.c.o
[ 23%] Building C object host/xtest/CMakeFiles/xtest.dir/adbg/src/adbg_expect.c.o
[ 26%] Building C object host/xtest/CMakeFiles/xtest.dir/adbg/src/adbg_log.c.o
[ 30%] Building C object host/xtest/CMakeFiles/xtest.dir/adbg/src/adbg_run.c.o
[ 33%] Building C object host/xtest/CMakeFiles/xtest.dir/adbg/src/security_utils_hex.c.o
[ 36%] Building C object host/xtest/CMakeFiles/xtest.dir/aes_perf.c.o
[ 40%] Building C object host/xtest/CMakeFiles/xtest.dir/benchmark_1000.c.o
[ 43%] Building C object host/xtest/CMakeFiles/xtest.dir/benchmark_2000.c.o
[ 46%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_1000.c.o
[ 50%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_4000.c.o
[ 53%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_4100.c.o
[ 56%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_5000.c.o
[ 60%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_6000.c.o
[ 63%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_7000.c.o
[ 66%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_8000.c.o
[ 70%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_8100.c.o
[ 73%] Building C object host/xtest/CMakeFiles/xtest.dir/sha_perf.c.o
[ 76%] Building C object host/xtest/CMakeFiles/xtest.dir/xtest_helpers.c.o
[ 80%] Building C object host/xtest/CMakeFiles/xtest.dir/xtest_main.c.o
[ 83%] Building C object host/xtest/CMakeFiles/xtest.dir/xtest_test.c.o
[ 86%] Building C object host/xtest/CMakeFiles/xtest.dir/regression_2000.c.o
[ 90%] Building C object host/xtest/CMakeFiles/xtest.dir/sock_server.c.o
[ 93%] Building C object host/xtest/CMakeFiles/xtest.dir/rand_stream.c.o
[ 96%] Building C object host/xtest/CMakeFiles/xtest.dir/install_ta.c.o
[100%] Linking C executable xtest
[100%] Built target xtest
```

**Comment by etienne-lms** [comment]
In my setup, my CMake home is not the optee_test but a parent dir where I cloned all optee_* repos and I made a **CMakeLitsts.txt**
```
add_subdirectory (optee_os)
add_subdirectory (optee_client)
...
```
From to parent dir I run (args removed) `cmake && make && make install`
With this patch build fails. Note that path from `CMAKE_HOME_DIRECTORY` seems an empty string:
```
make[2]: *** No rule to make target `cert/my.csr', needed by `optee_test/host/xtest/regression_8100_my_csr.h'.  Stop.
```
Maybe `CMAKE_HOME_DIRECTORY` is too restrictive on how cmake is used. Is there a variable that could be set from optee_test root  **CMakeLists.txt** to set the package source root path?

**Comment by etienne-lms** [comment]
:)  yes, many ways to build these packages, we cannot enforce one to use a specific setup 
The setup I described is mine and very minimal, but cmake could be used by various build env in various ways...

Looks like `CMAKE_CURRENT_SOURCE_DIR` sounds nice for optee_test top **CMakeLists.txt** to set the root path of the test source tree in a variable (`set OPTEE_TEST_ROOT_DIR ${CMAKE_CURRENT_SOURCE_DIR})`. Your throughs?

**Comment by etienne-lms** [comment]
The following appended to your change works fine with me (not yet tested the yocto way but should be ok).

```Patch
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,6 +4,8 @@ project (optee_test C)
 # Default cross compile settings
 set (CMAKE_TOOLCHAIN_FILE CMakeToolchain.txt)
 
+set (OPTEE_TEST_ROOT_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+
 ################################################################################
 # Compiler flags:
 #   We want to use the same flags in the entire optee_client git
--- a/host/xtest/CMakeLists.txt
+++ b/host/xtest/CMakeLists.txt
@@ -21,20 +21,20 @@ include(GNUInstallDirs)
 macro(EMBED_8100FILE prefix infile)
 	add_custom_command(
 		OUTPUT  regression_8100_${prefix}.h
-		COMMAND ${CMAKE_HOME_DIRECTORY}/scripts/file_to_c.py --inf ${infile}
+		COMMAND ${OPTEE_TEST_ROOT_DIR}/scripts/file_to_c.py --inf ${infile}
 	--out ${CMAKE_CURRENT_BINARY_DIR}/regression_8100_${prefix}.h
 	--name regression_8100_${prefix}
-		DEPENDS ${CMAKE_HOME_DIRECTORY}/scripts/file_to_c.py ${infile}
+		DEPENDS ${OPTEE_TEST_ROOT_DIR}/scripts/file_to_c.py ${infile}
 	)
 
 	set_property(SOURCE regression_8100.c APPEND PROPERTY OBJECT_DEPENDS
 		     ${CMAKE_CURRENT_BINARY_DIR}/regression_8100_${prefix}.h)
 endmacro(EMBED_8100FILE)
 
-EMBED_8100FILE(ca_crt ${CMAKE_HOME_DIRECTORY}/cert/ca.crt)
-EMBED_8100FILE(mid_crt ${CMAKE_HOME_DIRECTORY}/cert/mid.crt)
-EMBED_8100FILE(my_crt ${CMAKE_HOME_DIRECTORY}/cert/my.crt)
-EMBED_8100FILE(my_csr ${CMAKE_HOME_DIRECTORY}/cert/my.csr)
+EMBED_8100FILE(ca_crt ${OPTEE_TEST_ROOT_DIR}/cert/ca.crt)
+EMBED_8100FILE(mid_crt ${OPTEE_TEST_ROOT_DIR}/cert/mid.crt)
+EMBED_8100FILE(my_crt ${OPTEE_TEST_ROOT_DIR}/cert/my.crt)
+EMBED_8100FILE(my_csr ${OPTEE_TEST_ROOT_DIR}/cert/my.csr)
 
 set (SRC
 	adbg/src/adbg_case.c
@@ -95,7 +95,7 @@ add_executable (${PROJECT_NAME} ${SRC})
 target_compile_options (${PROJECT_NAME} PRIVATE -include conf.h)
 
 target_include_directories(${PROJECT_NAME}
-	PRIVATE .
+	PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}
 	PRIVATE adbg/include
 	PRIVATE xml/include
 	PRIVATE ${OPTEE_TEST_SDK}/host_include
```

**Comment by lorc** [comment]
Thank you, @etienne-lms, your approach is even more generic. Can I add your `Suggested-by`?

**Comment by etienne-lms** [comment]
If you add Joakim's one too :)    I mean you don't need to.

**Comment by jbech-linaro** [comment]
Nah, that's part of the review to give feedback and suggestions. But .. you may add my R-B instead.
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: host/xtest/CMakeLists.txt [hunk 1]
```diff
@@ -88,6 +88,7 @@ target_include_directories(${PROJECT_NAME}
 	PRIVATE adbg/include
 	PRIVATE xml/include
 	PRIVATE ${OPTEE_TEST_SDK}/host_include
+	PRIVATE ${CMAKE_CURRENT_BINARY_DIR}
```
Comment (etienne-lms) [inline host/xtest/CMakeLists.txt]: Why this change?

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

File: host/xtest/CMakeLists.txt (hunk 1 above)
Comment (lorc) [reply host/xtest/CMakeLists.txt]: Because of this parameter:

`--out ${CMAKE_CURRENT_BINARY_DIR}/regression_8100_${prefix}.h`

for `file_to_c.py` script.

I think, it should be in sync. 

Also, it is not building without this change.

**Comment by lorc** [comment]
Thanks. 

I updated the patch, as Joakim and  Etienne suggested and reworked commit message accordingly.

Tags are updated also.

File: host/xtest/CMakeLists.txt (hunk 1 above)
Comment (etienne-lms) [reply host/xtest/CMakeLists.txt]: ok, thanks.
